### PR TITLE
feat(hummock manager): let client get version delta rather than versi…

### PR DESCRIPTION
**Pull request by soundOfDestiny**
…on in `pin_version`

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

let client get version delta rather than version in pin_version
see #3597 

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

File: src/storage/src/hummock/local_version_manager.rs [hunk 1]
```diff
@@ -213,16 +215,33 @@ impl LocalVersionManager {
...
+        pin_resp: (bool, Vec<HummockVersionDelta>, HummockVersion),
+    ) -> bool {
+        let old_version = self.local_version.upgradable_read();
+        if let Some(last_pinned_id) = last_pinned {
+            if old_version.pinned_version().id() != last_pinned_id {
+                return false;
+            }
+        }
+        let newly_pinned_version = if pin_resp.0 {
+            let mut version_to_apply = old_version.pinned_version().version();
```
Comment (zwang28) [inline src/storage/src/hummock/local_version_manager.rs:230]: Seems this line has introduced a constant overhead of version copy?
We can defer this cost until seeing old_version.pinned_version().id() < new_version_id.

File: src/meta/src/rpc/service/hummock_service.rs [hunk 2]
```diff
@@ -77,10 +77,14 @@ where
...
-            Ok(pinned_version) => Ok(Response::new(PinVersionResponse {
-                status: None,
-                pinned_version: Some(pinned_version),
-            })),
+            Ok((is_delta_response, version_deltas, pinned_version)) => {
+                Ok(Response::new(PinVersionResponse {
+                    status: None,
+                    is_delta_response,
+                    version_deltas,
+                    pinned_version: Some(pinned_version),
```
[the remark this answers is not in the document]
Comment (soundOfDestiny) [reply src/meta/src/rpc/service/hummock_service.rs]: > 

fixed

File: src/storage/src/hummock/local_version_manager.rs (hunk 1 above)
Comment (soundOfDestiny) [reply src/storage/src/hummock/local_version_manager.rs:244]: > Seems this line has introduced a constant overhead of version copy, even when old_version.pinned_version().id() >= new_version_id.

fixed

File: src/storage/src/hummock/local_version.rs [hunk 3]
```diff
@@ -169,7 +169,6 @@ impl PinnedVersion {
         self.version.safe_epoch
     }
 
-    #[cfg(any(test, feature = "test"))]
```
[the remark this answers is not in the document]
Comment (soundOfDestiny) [reply src/storage/src/hummock/local_version.rs:172]: > The version method itself is dangerous if used without caution by others. This HummockVersion may have been unpinned after PinnedVersion is dropped.
> 
> What about leaving a comment to warn it, or any other suggestions?

just added some comment

**Review by zwang28 — APPROVED** [review]
LGTM

**Pull request merged**